libbeat/common/dtfmt/dtfmt_test.go

```diff
@@ -82,10 +82,37 @@ func TestFormat(t *testing.T) {
 		{mkTime(8, 5, 24, 0), "kk:mm:ss aa", "09:05:24 AM"},
 		{mkTime(20, 5, 24, 0), "k:m:s a", "21:5:24 PM"},
 		{mkTime(20, 5, 24, 0), "kk:mm:ss aa", "21:05:24 PM"},
-		{mkTime(1, 2, 3, 123), "S", "1"},
-		{mkTime(1, 2, 3, 123), "SS", "12"},
-		{mkTime(1, 2, 3, 123), "SSS", "123"},
-		{mkTime(1, 2, 3, 123), "SSSS", "1230"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "S", "1"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "SS", "12"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "SSS", "123"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "SSSS", "1230"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "f", "1"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "ff", "12"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "fff", "123"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "ffff", "123"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "fffff", "123"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "ffffff", "123"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "fffffff", "123"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "ffffffff", "123"},
+		{mkTime(1, 2, 3, 123*time.Millisecond), "fffffffff", "123"},
+		{mkTime(1, 2, 3, 123*time.Microsecond), "f", "0"},
+		{mkTime(1, 2, 3, 123*time.Microsecond), "ff", "00"},
+		{mkTime(1, 2, 3, 123*time.Microsecond), "fff", "000"},
+		{mkTime(1, 2, 3, 123*time.Microsecond), "ffff", "0001"},
+		{mkTime(1, 2, 3, 123*time.Microsecond), "fffff", "00012"},
+		{mkTime(1, 2, 3, 123*time.Microsecond), "ffffff", "000123"},
+		{mkTime(1, 2, 3, 123*time.Microsecond), "fffffff", "000123"},
+		{mkTime(1, 2, 3, 123*time.Microsecond), "ffffffff", "000123"},
+		{mkTime(1, 2, 3, 123*time.Microsecond), "fffffffff", "000123"},
```

Comment:
Just curious: Why are the trailing zeros removed? As a user, I would expect to see as many digits as I configured in the format string.

Reply:
We have multiple formatters for fractions of a second: `S`, `n`, and `f`:
- `S` (fraction of second): print as many digits as set in format string
- `n` (nanoseconds): always print 9 digits independent of the number of `n` in the format string
- `f` (precision aware fraction of second): if timestamp precision is millisecond we print 3, if microsecond we print 6, and if nanosecond we print 9 digits. 

By using `f` in Beats we would keep the original timestamp precision in case we parse a timestamp from a JSON file.